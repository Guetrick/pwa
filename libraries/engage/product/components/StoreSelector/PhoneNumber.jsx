import React, { memo } from 'react';
import PropTypes from 'prop-types';
import PhoneIcon from '@shopgate/pwa-ui-shared/icons/PhoneIcon';
import { phone as phoneStyle, phoneIcon } from './style';

/**
 * Renders the store's phone number.
 * @param {Object} props The compnent props.
 * @returns {JSX}
 */
const PhoneNumber = ({ phone }) => {
  if (!phone) {
    return null;
  }

  return (
    <a href={`tel:${phone}`} className={phoneStyle}>
      <span className={phoneIcon}>
        <PhoneIcon />
      </span>
      <span>{phone}</span>
    </a>
  );
};

PhoneNumber.propTypes = {
  phone: PropTypes.string,
};

PhoneNumber.defaultProps = {
  phone: null,
};

export default memo(PhoneNumber);
