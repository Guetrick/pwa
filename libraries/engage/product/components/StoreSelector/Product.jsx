import React, { memo, useContext } from 'react';
import Ellipsis from '@shopgate/pwa-common/components/Ellipsis';
import { ProductImage } from '../../index';
import StoreSelectorContext from './context';
import * as styles from './style';

/**
 * Renders the product information.
 * @returns {JSX}
 */
const Product = () => {
  const { product, selectedVariants } = useContext(StoreSelectorContext);

  if (!product) {
    return null;
  }

  return (
    <div className={styles.productContainer}>
      <div className={styles.productContainerInner}>
        <div className={styles.productImage}>
          <ProductImage src={product.featuredImageUrl} />
        </div>
        <div className={styles.productContent}>
          <div className={styles.productName}>
            <Ellipsis>{product.name}</Ellipsis>
          </div>
          {selectedVariants.length > 0 && (
            <div className={styles.variants}>
              {selectedVariants.map(({ label, value }) => (
                <div key={`${label}-${value}`}>{`${label}: ${value}`}</div>
              ))}
            </div>
          )}
          <div className={styles.availability}>
            <div className={styles.availabilityText}>Ready for pick up in 1-3 days in</div>
            <div className={styles.availabilityStores}>13 Stores</div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default memo(Product);
