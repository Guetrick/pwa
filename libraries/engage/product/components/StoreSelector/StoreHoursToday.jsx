import React, { memo } from 'react';
import PropTypes from 'prop-types';
import I18n from '@shopgate/pwa-common/components/I18n';
import { storeHoursToday } from './style';

const weekdays = ['sun', 'mon', 'tue', 'wed', 'thu', 'fri', 'sat'];

/**
 * Renders the store's opening hours for "today".
 * @param {Object} props The component props.
 * @returns {JSX}
 */
const StoreHoursToday = ({ hours }) => {
  const today = weekdays[new Date().getDay()];
  const hoursToday = hours[today] || null;

  if (!hoursToday || !hoursToday === '') {
    return null;
  }

  return (
    <div className={storeHoursToday}>
      <I18n.Text string="product.location.today_hours" params={{ hours: hoursToday }} />
    </div>
  );
};

StoreHoursToday.propTypes = {
  hours: PropTypes.shape().isRequired,
};

export default memo(StoreHoursToday);
