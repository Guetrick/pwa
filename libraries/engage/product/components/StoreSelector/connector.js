import { connect } from 'react-redux';
import { makeGetProductLocations, getProduct } from '@shopgate/pwa-common-commerce/product';
import { selectLocation } from '@shopgate/pwa-common/action-creators';
import { makeGetSelectedCharacteristic } from '../../selectors/variants';

/**
 * @returns {Function}
 */
function makeMapStateToProps() {
  const getProductLocations = makeGetProductLocations();
  const getSelectedCharacteristic = makeGetSelectedCharacteristic();

  /**
   * @param {Object} state The application state.
   * @param {Object} props The component props.
   * @returns {Object}
   */
  return (state, props) => ({
    locations: getProductLocations(state, props),
    product: getProduct(state, props),
    selectedVariants: getSelectedCharacteristic(state, props),
  });
}

/**
 * @param {Function} dispatch The dispatch function.
 * @returns {Object}
 */
const mapDispatchToProps = dispatch => ({
  selectLocation: location => dispatch(selectLocation(location)),
});

export default connect(makeMapStateToProps, mapDispatchToProps);
