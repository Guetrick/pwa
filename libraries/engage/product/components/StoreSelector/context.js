import { createContext } from 'react';

const StoreSelectorContext = createContext();

export default StoreSelectorContext;
