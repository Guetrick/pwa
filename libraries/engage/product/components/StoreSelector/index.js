export { default as StoreSelector } from './StoreSelector';
export { default as StoreSelectorContext } from './context';
