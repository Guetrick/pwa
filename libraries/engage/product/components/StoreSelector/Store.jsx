import React, { memo, useCallback, useContext } from 'react';
import PropTypes from 'prop-types';
import { Accordion } from '@shopgate/pwa-ui-material';
import I18n from '@shopgate/pwa-common/components/I18n';
import Address from './Address';
import StoreHoursToday from './StoreHoursToday';
import PhoneNumber from './PhoneNumber';
import OpeningHours from './OpeningHours';
import { store as storeStyles, storeMain, storeImage, storeName, storeInfo } from './style';
import StoreSelectorContext from './context';

/**
 * Renders the store's addresses as pickup locations.
 * @param {Object} props The component props.
 * @returns {JSX}
 */
const Store = ({ store }) => {
  const { selectLocation, close } = useContext(StoreSelectorContext);

  if (!store || !store.addresses || store.addresses.length === 0) {
    return null;
  }

  return store.addresses.map((address) => {
    const handleClick = useCallback((event) => {
      event.preventDefault();
      selectLocation({
        locationCode: store.code,
        addressCode: address.code,
        visibleStock: store.inventory.visible,
      });
      close();
    }, [address, store]);

    return (
      <div className={storeStyles} key={address.code}>
        <div className={storeMain}>
          <div className={storeImage} />
          <div
            className={storeInfo}
            onClick={handleClick}
            onKeyDown={handleClick}
            role="button"
            tabIndex={0}
          >
            <div className={storeName}>{store.name}</div>
            <StoreHoursToday hours={store.operationHours} />
            <Address address={address} />
          </div>
        </div>
        <Accordion renderLabel={() => <I18n.Text string="product.location.details" />}>
          <OpeningHours hours={store.operationHours} />
          <PhoneNumber phone={address.phoneNumber} />
        </Accordion>
      </div>
    );
  });
};

Store.propTypes = {
  store: PropTypes.shape(),
};

Store.defaultProps = {
  store: null,
};

export default memo(Store);
