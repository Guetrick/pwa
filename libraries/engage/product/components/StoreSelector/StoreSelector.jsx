import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { UIEvents, withCurrentProduct } from '../../../core';
import { SheetDrawer } from '../../../components';
import { sheet } from './style';
import Product from './Product';
import Locations from './Locations';
import connect from './connector';
import StoreSelectorContext from './context';

const EVENT_SET_OPEN = 'event.setOpen';
const EVENT_SET_CLOSED = 'event.setClosed';

/**
 * Renders the store selector sheet.
 */
class StoreSelector extends PureComponent {
  /**
   * Opens the store selector sheet.
   */
  static open = () => {
    UIEvents.trigger(EVENT_SET_OPEN);
  }

  /**
   * Closes the store selector sheet.
   */
  static close = () => {
    UIEvents.trigger(EVENT_SET_CLOSED);
  }

  static propTypes = {
    locations: PropTypes.arrayOf(PropTypes.shape()),
    product: PropTypes.shape(),
    selectedVariants: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.string,
      value: PropTypes.string,
    })),
    selectLocation: PropTypes.func,
  }

  static defaultProps = {
    locations: null,
    product: null,
    selectedVariants: [],
    selectLocation: () => { },
  }

  static contextTypes = {
    i18n: PropTypes.func,
  };

  /**
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);

    UIEvents.addListener(EVENT_SET_OPEN, this.handleSetOpen);
    UIEvents.addListener(EVENT_SET_CLOSED, this.handleSetOpen);
  }

  state = {
    isOpen: false,
  };

  /**
   * Deregisters the event listeners.
   */
  componentWillUnmount() {
    UIEvents.removeListener(EVENT_SET_OPEN, this.handleSetOpen);
    UIEvents.removeListener(EVENT_SET_CLOSED, this.handleSetOpen);
  }

  handleSetOpen = () => {
    this.setState({ isOpen: true });
  }

  handleSetClosed = () => {
    this.setState({ isOpen: false });
  }

  /**
   * Handles the selection of a store location from the sheet.
   * @param {Object} params The parameters to store.
   * @param {string} params.locationCode The location code.
   * @param {string} params.addresscode The address code.
   * @param {number} params.visibleStock The visible stock amount.
   */
  handleSelectLocation = ({ locationCode, addressCode, visibleStock }) => {
    const { selectLocation, product } = this.props;
    const location = {
      productCode: product.id,
      locationCode,
      addressCode,
      visibleStock,
    };

    selectLocation(location);
  }

  /**
   * @returns {JSX}
   */
  render() {
    const { product, locations, selectedVariants } = this.props;
    const { isOpen } = this.state;
    const { __ } = this.context.i18n();

    const contextValue = {
      selectLocation: this.handleSelectLocation,
      close: this.handleSetClosed,
      product,
      locations,
      selectedVariants,
    };

    return (
      <StoreSelectorContext.Provider value={contextValue}>
        <SheetDrawer
          isOpen={isOpen}
          title={__('product.location.headline')}
          onClose={this.handleSetClosed}
        >
          <div className={sheet}>
            <Product />
            <Locations />
          </div>
        </SheetDrawer>
      </StoreSelectorContext.Provider>
    );
  }
}

export default withCurrentProduct(connect(StoreSelector));
