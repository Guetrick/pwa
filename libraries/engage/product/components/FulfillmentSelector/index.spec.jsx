import React from 'react';
import { shallow } from 'enzyme';
import {
  PRODUCT_FULFILLMENT_METHOD_DIRECT_SHIP,
  PRODUCT_FULFILLMENT_METHOD_IN_STORE_PICKUP,
} from '../../constants';
import { StoreSelector } from '../StoreSelector';
import { FulfillmentSelector } from './';

jest.mock('../../../core/config/isBeta', () => ({
  isBeta: jest.fn().mockReturnValue(true),
}));
jest.mock('../StoreSelector', () => ({
  StoreSelector: {
    open: jest.fn(),
  },
}));

describe('<FulfillmentSelector />', () => {
  it('should not render if the fulfillmentMethods do not contain inStorePickup', () => {
    const wrapper = shallow(<FulfillmentSelector
      productCode="abc-123"
      location={{
        name: 'Test Store',
        visibleInventory: 15,
      }}
      fulfillmentMethods={[]}
    />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.instance()).toEqual(null);
  });

  it('should render the fulfillment selector with two radio elements', () => {
    const wrapper = shallow(<FulfillmentSelector
      productCode="abc-123"
      location={{
        name: 'Test Store',
        visibleInventory: 15,
      }}
      fulfillmentMethods={
        [PRODUCT_FULFILLMENT_METHOD_DIRECT_SHIP, PRODUCT_FULFILLMENT_METHOD_IN_STORE_PICKUP]
      }
    />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('RadioItem').length).toBe(2);
    expect(wrapper.find('[name="product.fulfillment_selector.direct_ship"]').length).toBe(1);
    expect(wrapper.find('[name="product.fulfillment_selector.pick_up_in_store"]').length).toBe(1);
  });

  it('should open the store selector sheet on selecting pick_up_in_store', () => {
    const wrapper = shallow(<FulfillmentSelector
      productCode="abc-123"
      location={{
        name: 'Test Store',
        visibleInventory: 15,
      }}
      fulfillmentMethods={
        [PRODUCT_FULFILLMENT_METHOD_DIRECT_SHIP, PRODUCT_FULFILLMENT_METHOD_IN_STORE_PICKUP]
      }
    />);
    wrapper.find('[name="product.fulfillment_selector"]')
      .simulate('change', 'product.fulfillment_selector.pick_up_in_store');
    expect(StoreSelector.open).toHaveBeenCalledTimes(1);
  });
});
