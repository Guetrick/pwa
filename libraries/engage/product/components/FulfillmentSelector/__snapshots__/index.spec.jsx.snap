// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FulfillmentSelector /> should not render if the fulfillmentMethods do not contain inStorePickup 1`] = `""`;

exports[`<FulfillmentSelector /> should open the store selector sheet on selecting pick_up_in_store 1`] = `
<SurroundPortals
  portalName="product.fulfillment-selector"
  portalProps={
    Object {
      "productCode": "abc-123",
    }
  }
>
  <div
    className="css-12ywm8s"
    data-test-id="product-fulfillment-selector"
  >
    <div
      aria-hidden={true}
      className="css-10lhp8m"
      role="heading"
    >
      <Translate
        className=""
        params={Object {}}
        role={null}
        string="product.fulfillment_selector.heading"
        transform={null}
      />
    </div>
    <RadioGroup
      className="css-n7bvec"
      direction="column"
      errorText=""
      isControlled={true}
      label=""
      name="product.fulfillment_selector"
      onChange={[Function]}
      translateErrorText={true}
      value="product.fulfillment_selector.direct_ship"
    >
      <RadioItem
        checked={false}
        className="css-i8mws5"
        label={
          <div
            className="css-1hju4sd css-1siwvsy"
          >
            <Translate
              className=""
              params={Object {}}
              role={null}
              string="product.fulfillment_selector.direct_ship"
              transform={null}
            />
            <Connect(pure(Availability))
              fulfillmentSelection="directShip"
              productId="abc-123"
            />
          </div>
        }
        name="product.fulfillment_selector.direct_ship"
        onChange={[Function]}
      />
      <RadioItem
        checked={false}
        className="css-i8mws5"
        label={
          <div
            className="css-1hju4sd css-1siwvsy"
          >
            <Translate
              className=""
              params={Object {}}
              role={null}
              string="product.fulfillment_selector.pick_up_in_store"
              transform={null}
            />
            <pure(LocationStockInfo)
              location={
                Object {
                  "name": "Test Store",
                  "visibleInventory": 15,
                }
              }
            />
          </div>
        }
        name="product.fulfillment_selector.pick_up_in_store"
        onChange={[Function]}
      />
    </RadioGroup>
  </div>
</SurroundPortals>
`;

exports[`<FulfillmentSelector /> should render the fulfillment selector with two radio elements 1`] = `
<SurroundPortals
  portalName="product.fulfillment-selector"
  portalProps={
    Object {
      "productCode": "abc-123",
    }
  }
>
  <div
    className="css-12ywm8s"
    data-test-id="product-fulfillment-selector"
  >
    <div
      aria-hidden={true}
      className="css-10lhp8m"
      role="heading"
    >
      <Translate
        className=""
        params={Object {}}
        role={null}
        string="product.fulfillment_selector.heading"
        transform={null}
      />
    </div>
    <RadioGroup
      className="css-n7bvec"
      direction="column"
      errorText=""
      isControlled={true}
      label=""
      name="product.fulfillment_selector"
      onChange={[Function]}
      translateErrorText={true}
      value="product.fulfillment_selector.direct_ship"
    >
      <RadioItem
        checked={false}
        className="css-i8mws5"
        label={
          <div
            className="css-1hju4sd css-1siwvsy"
          >
            <Translate
              className=""
              params={Object {}}
              role={null}
              string="product.fulfillment_selector.direct_ship"
              transform={null}
            />
            <Connect(pure(Availability))
              fulfillmentSelection="directShip"
              productId="abc-123"
            />
          </div>
        }
        name="product.fulfillment_selector.direct_ship"
        onChange={[Function]}
      />
      <RadioItem
        checked={false}
        className="css-i8mws5"
        label={
          <div
            className="css-1hju4sd css-1siwvsy"
          >
            <Translate
              className=""
              params={Object {}}
              role={null}
              string="product.fulfillment_selector.pick_up_in_store"
              transform={null}
            />
            <pure(LocationStockInfo)
              location={
                Object {
                  "name": "Test Store",
                  "visibleInventory": 15,
                }
              }
            />
          </div>
        }
        name="product.fulfillment_selector.pick_up_in_store"
        onChange={[Function]}
      />
    </RadioGroup>
  </div>
</SurroundPortals>
`;
